lib/fileset/internal.nix

```diff
@@ -638,4 +638,66 @@ rec {
     else
       # In all other cases it's the rhs
       rhs;
+
+  # Compute the set difference between two file sets.
+  # The filesets must already be coerced and validated to be in the same filesystem root.
+  # Type: Fileset -> Fileset -> Fileset
+  _difference = fileset1: fileset2:
+    let
+      # The common base components prefix, e.g.
+      # (/foo/bar, /foo/bar/baz) -> /foo/bar
+      # (/foo/bar, /foo/baz) -> /foo
+      commonBaseComponentsLength =
+        # TODO: Have a `lib.lists.commonPrefixLength` function such that we don't need the list allocation from commonPrefix here
+        length (
+          commonPrefix
+            fileset1._internalBaseComponents
+            fileset2._internalBaseComponents
+        );
+
+      # The filesetTree's need to have the same base path in order to be able to compute the difference between them.
```

Comment:
```suggestion
      # The filesetTree's need to have the same base path in order to be able to compute the difference between them.
      # Technically an unrelated root in the second, negative argument could be considered not to match any files in
      # the first argument, but users would be clueless as to why their filter doesn't work.
```

Reply:
Hmm not 100% sure what you mean here. If the root of the second argument is unrelated (e.g. `difference /foo /bar`), then no files inside `/foo` match `/bar`, meaning nothing gets subtracted, so the result is `/foo`. And that should be expected.

Reply:
I was thinking of entirely unrelated trees, such as when one is an absolute path, but the other is a `fetchGit` lazy tree.

Rejecting `difference /foo /bar` seems excessive, I agree.

Reply:
Ah yeah, it will error for different filesystem roots, there's a comment on this in the comments for `_difference`:

```
# The filesets must already be coerced and validated to be in the same filesystem root.
```

The public `difference` function passes its arguments through `_coerceMany`, which checks against this, before calling the internal one.